File: frontend/src/nodes2/nodes/krea2_tiled_diffusion.ts
import { defineAsyncComponent } from "vue";
import type { ComfyNodeData, LGraphNode, LGraphNodeType } from "@/types/comfy";
import type { NodeModule } from "@/nodes2/nodeRegistry";
import { registerStyledNode } from "@/nodes2/nodeStyle";
import { addFilDomWidget, unmountAllFilWidgets } from "@/nodes2/domWidgetHost";
import { createSyncedNodeState, findFilWidget, hideNativeWidget, sanitizeWidgetValue } from "@/nodes2/util";
import { exposeWidgetInputSockets, installWidgetSocketSync } from "@/nodes2/widgetInputSockets";
import { applyFxComposables } from "@/nodes2/applyFxComposables";
import { FIL_STATE_KEY, installFilStatePersistence, restoreFilState, type PersistedPanelState } from "@/nodes2/statePersistence";

const Krea2Vue = defineAsyncComponent(() => import("@/components/nodes/Krea2TiledDiffusionPanel.vue"));

export const KREA2_SOCKET_INPUTS = [
  "prompt",
  "seed",
  "steps",
  "denoise",
  "vision_weight",
  "upscale_factor",
  "tile_grid",
  "tile_overlap",
  "tile_batch_size",
  "texture_injection",
  "color_match",
  "identity_lora_name",
  "identity_lora_strength",
];

const numericDefaults: Record<string, number> = {
  seed: 0,
  steps: 8,
  denoise: 0.22,
  vision_weight: 1.40,
  upscale_factor: 2.0,
  tile_batch_size: 1,
  texture_injection: 0.20,
  identity_lora_strength: 1.0,
};

const stringDefaults: Record<string, string> = {
  prompt: "hyperrealistic, highly detailed, 8k uhd",
  tile_grid: "auto",
  tile_overlap: "auto (256px)",
  color_match: "none",
  identity_lora_name: "none",
  control_after_generate: "randomize",
};

const HIDE = [...Object.keys(numericDefaults), ...Object.keys(stringDefaults)];

export const krea2TiledDiffusionNode: NodeModule = {
  id: "FiLKrea2TiledDiffusion",
  register(nodeType: LGraphNodeType, _nodeData: ComfyNodeData): void {
    registerStyledNode(nodeType, {
      minSize: [320, 560],
      initialWidth: 320,
      family: "image",
      description: "Ultra-high-definition tiled upscale with RoPE canvas coordinates, edge-aware adaptive texture and color matching.",
      badges: [{ text: "krea2", color: "#62c987", text_color: "#1a1a1a" }],
    });

    const proto = nodeType as {
      prototype: {
        onNodeCreated?: (...a: unknown[]) => unknown;
        onConfigure?: (...a: unknown[]) => unknown;
        onRemoved?: (...a: unknown[]) => unknown;
      };
    };
    const p = proto.prototype;

    const syncAll = (node: LGraphNode, target: Record<string, unknown>, quiet = false) => {
      for (const name of Object.keys(numericDefaults)) {
        target[name] = sanitizeWidgetValue(findFilWidget(node, name), "number", numericDefaults[name], quiet);
      }
      for (const name of Object.keys(stringDefaults)) {
        target[name] = sanitizeWidgetValue(findFilWidget(node, name), "string", stringDefaults[name], quiet);
      }
    };

    const originalCreated = p.onNodeCreated;
    p.onNodeCreated = function (this: LGraphNode, ...args: unknown[]) {
      const result = originalCreated?.apply(this, args);
      const node = this as LGraphNode & { _filKrea2State?: unknown; _filSocketPolicy?: string };
      node._filSocketPolicy = "always";
      const initial: Record<string, unknown> = {};
      syncAll(node, initial);
      for (const name of HIDE) {
        hideNativeWidget(node, name);
      }
      const state = {
        nodeState: createSyncedNodeState(node, initial),
        initialValues: { ...initial },
        ui: {},
      };
      Object.defineProperty(state, "node", { value: node, enumerable: false, configurable: true });
      node._filKrea2State = state;

      installFilStatePersistence(node, state);
      addFilDomWidget(node, "fil_krea2_view", Krea2Vue, { state, height: 560, growable: true });
      exposeWidgetInputSockets(this, KREA2_SOCKET_INPUTS);
      return result;
    };

    const originalConfigure = p.onConfigure;
    p.onConfigure = function (this: LGraphNode, ...args: unknown[]) {
      const result = originalConfigure?.apply(this, args);
      const node = this as LGraphNode & { _filKrea2State?: PersistedPanelState; _filSocketPolicy?: string };
      node._filSocketPolicy = "always";
      const state = node._filKrea2State;
      if (!state) return result;
      const hasFilState = Boolean((args[0] as Record<string, unknown> | undefined)?.[FIL_STATE_KEY]);
      syncAll(node, state.nodeState, hasFilState);
      restoreFilState(state, args[0]);
      exposeWidgetInputSockets(this, KREA2_SOCKET_INPUTS);
      return result;
    };

    const originalRemoved = p.onRemoved;
    p.onRemoved = function (this: LGraphNode, ...args: unknown[]) {
      unmountAllFilWidgets(this);
      return originalRemoved?.apply(this, args);
    };

    installWidgetSocketSync(p, KREA2_SOCKET_INPUTS, "_filKrea2State");
    applyFxComposables(nodeType as { prototype?: unknown });
  },
};

File: common/contracts/nodes/krea2_tiled_diffusion.py
"""Frontend contract for `FiLKrea2TiledDiffusion`."""

from __future__ import annotations

from ..widgets import _combo, _float, _int, _slider, _string
from ..schema import NodeContract, NodeInputs, NodeOutput
from ...brand import CATEGORY_IMAGE
TILE_GRIDS = ["1x1", "1x2", "2x1", "2x2", "2x3", "3x2", "3x3", "4x4"]
ALL_TILE_GRIDS = ["auto"] + TILE_GRIDS
TILE_OVERLAP_OPTIONS = ["auto (256px)", "128px", "192px", "256px", "384px", "512px"]
COLOR_MATCH_OPTIONS = ["none", "luminance", "wavelet"]

CONTRACT = NodeContract(
    id="FiLKrea2TiledDiffusion",
    title="📐 Krea 2 Tiled Diffusion",
    category=CATEGORY_IMAGE,
    description="Ultra-high-definition tiled upscale with RoPE canvas coordinates, edge-aware adaptive texture and color matching.",
    min_size=(320, 560),
    family="image",
    inputs=NodeInputs(
        required=[
            _string("prompt", default="hyperrealistic, highly detailed, 8k uhd", label="Prompt"),
            _int("seed", default=0, minv=0, maxv=0xFFFFFFFFFFFFFFFF, step=1, label="Seed"),
            _int("steps", default=8, minv=1, maxv=50, step=1, label="Steps"),
            _slider("denoise", default=0.22, minv=0.0, maxv=1.0, step=0.01, label="Denoise"),
            _float("vision_weight", default=1.40, minv=0.0, maxv=3.0, step=0.05, label="Vision weight"),
            _slider("upscale_factor", default=2.0, minv=1.0, maxv=8.0, step=0.1, label="Upscale factor"),
            _combo("tile_grid", values=ALL_TILE_GRIDS, default="auto", label="Tile grid"),
            _combo("tile_overlap", values=TILE_OVERLAP_OPTIONS, default="auto (256px)", label="Tile overlap"),
            _int("tile_batch_size", default=1, minv=1, maxv=8, step=1, label="Tile batch size"),
            _slider("texture_injection", default=0.20, minv=0.0, maxv=1.0, step=0.02, label="Texture injection"),
            _combo("color_match", values=COLOR_MATCH_OPTIONS, default="none", label="Color match"),
            _combo("identity_lora_name", values=["none"], default="none", label="Identity LoRA"),
            _float("identity_lora_strength", default=1.0, minv=0.0, maxv=2.0, step=0.05, label="LoRA strength"),
        ],
        optional=[],
    ),
    outputs=[
        NodeOutput(name="image", type="IMAGE"),
        NodeOutput(name="latent", type="LATENT"),
    ],
)
